url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/utils/StatisticalSummary.java

```diff
@@ -0,0 +1,327 @@
+        return sumsq - n * mean() * mean();
+    }
+
+    private void computeStats() {
+        if (!valid) {
+            mean = sum / n;
+            gm = Math.exp(logsum / n);
+            double num = sumsq - (n * mean * mean);
+            if (num < 0) {
+                // avoids tiny negative numbers possible through imprecision
+                num = 0;
+            }
+            // System.out.println("Num = " + num);
+            sd = Math.sqrt(num / (n - 1));
+            // System.out.println(" Test: sd = " + sd);
+            // System.out.println(" Test: n = " + n);
+            valid = true;
+        }
+    }
+
+    public double sd() {
+        if (!valid)
+            computeStats();
+        return sd;
+    }
+
+    public int n() {
+        return n;
+    }
+
+    public double stdErr() {
+        return sd() / Math.sqrt(n);
+    }
+
+    public void add(StatisticalSummary ss) {
+        // implications for Watch?
+        n += ss.n;
+        sum += ss.sum;
+        sumsq += ss.sumsq;
+        logsum += ss.logsum;
+        max = Math.max(max, ss.max);
+        min = Math.min(min, ss.min);
+        valid = false;
+    }
+
+    public void add(double d) {
+        n++;
+        sum += d;
+        sumsq += d * d;
+        if (d > 0) {
+            logsum += Math.log(d);
+        }
+        min = Math.min(min, d);
+        max = Math.max(max, d);
+        if (watch != null) {
+            watch.note(d);
+        }
+        valid = false;
+    }
+
+    public void add(Number n) {
+        add(n.doubleValue());
+    }
+
+    public void add(double[] d) {
+        for (int i = 0; i < d.length; i++) {
+            add(d[i]);
+        }
+    }
+
+    public void add(Vector v) {
+        for (int i = 0; i < v.size(); i++) {
+            try {
+                add(((Number) v.elementAt(i)).doubleValue());
+            } catch (Exception e) {
+            }
+        }
+    }
+
+    public String toString() {
+        String s = (name == null) ? "" : name + "\n";
+        s += " min = " + min() + "\n" +
+                " max = " + max() + "\n" +
+                " ave = " + mean() + "\n" +
+                " sd  = " + sd() + "\n" +
+               // " se  = " + stdErr() + "\n" +
+               // " sum  = " + sum + "\n" +
+               // " sumsq  = " + sumsq + "\n" +
+               // " watch = " + watch + "\n" +
+                " n   = " + n;
+        return s;
+
+    }
+
+    public void save(String path) {
+        try {
+            ObjectOutputStream oos =
+                    new ObjectOutputStream(
+                            new FileOutputStream(path));
+            oos.writeObject(this);
+            oos.close();
+        } catch (Exception e) {
+            System.out.println(e);
+        }
+    }
+
+    public static StatisticalSummary load(String path) {
+        try {
+            ObjectInputStream ois =
+                    new ObjectInputStream(
+                            new FileInputStream(path));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. The resource is referenced at lines 256, 258. The resource is closed at line 259. There are other execution paths that do not contain closure statements, for example, when ObjectInputStream.readObject throws an exception. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).